src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1911 @@
+                        "--json=short",
+                        NULL, /* maybe --verify=no */
+                        NULL, /* maybe --component=, --root=, or --image= */
+                        NULL, /* maybe --offline */
+                        NULL, /* list, check-new, update, vacuum */
+                        NULL, /* maybe version (for list, update) */
+                        NULL
+                };
+                unsigned k = 2;
+
+                if (setenv("NOTIFY_SOCKET", "/run/systemd/sysupdate/notify", 1) < 0) {
+                        log_error_errno(errno, "setenv() failed: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                if (getenv_bool("SYSTEMD_SYSUPDATE_NO_VERIFY") > 0)
+                        cmd[k++] = "--verify=no"; /* For testing */
+
+                r = setenv_systemd_exec_pid(true);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to update $SYSTEMD_EXEC_PID, ignoring: %m");
+
+                r = target_get_argument(j->target, &target_arg);
+                if (r < 0) {
+                        log_oom();
+                        _exit(EXIT_FAILURE);
+                }
+                if (target_arg)
+                        cmd[k++] = target_arg;
+
+                if (j->offline)
+                        cmd[k++] = "--offline";
+
+                switch (j->type) {
+                case JOB_LIST:
+                        cmd[k++] = "list";
+                        break;
+
+                case JOB_DESCRIBE:
+                        cmd[k++] = "list";
+                        assert(!isempty(j->version));
+                        cmd[k++] = j->version;
+                        break;
+
+                case JOB_CHECK_NEW:
+                        cmd[k++] = "check-new";
+                        break;
+
+                case JOB_UPDATE:
+                        cmd[k++] = "update";
+                        cmd[k++] = empty_to_null(j->version);
+                        break;
+
+                case JOB_VACUUM:
+                        cmd[k++] = "vacuum";
+                        break;
+
+                default:
+                        assert_not_reached();
+                }
+
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *s = NULL;
+
+                        s = quote_command_line((char**) cmd, SHELL_ESCAPE_EMPTY);
+                        if (!s) {
+                                log_oom();
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        log_debug("Spawning worker for job %" PRIu64 ": %s", j->id, s);
+                }
+
+                r = invoke_callout_binary(sysupdate_binary_path(), (char *const *) cmd);
+                log_error_errno(r, "Failed to execute systemd-sysupdate: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        r = sd_event_add_child(j->manager->event, &j->child, pid, WEXITED, job_on_exit, j);
```

Comment:
here too, for new stuff let's use event_add_child_pidref() right from the beginning.